### PR TITLE
vam: Add grok()

### PR DESCRIPTION
**Pull request by mattnibs**

### CODE REVIEW

File: runtime/ztests/expr/function/grok.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
+zed: yield grok(pattern, input, defs)
+
+vector: true
+
+input: |
+  {
+    pattern: "%{TIMESTAMP_ISO8601:event_time} %{LOGLEVEL:log_level} %{GREEDYDATA:log_message}",
+    input: "2020-09-16T04:20:42.45+01:00 DEBUG This is a sample debug log message",
+    defs: null(string)
+  }
```
Comment (nwt) [inline runtime/ztests/expr/function/grok.yaml:10]: While you're here, maybe add test cases for null(string) pattern and input along with non-string pattern, input, and defs?

File: runtime/vam/expr/function/grok.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+func newGrok(zctx *super.Context) *Grok {
+	return &Grok{
+		zctx:  zctx,
+		hosts: make(map[string]*host),
+	}
+}
+
+func (g *Grok) Call(args ...vector.Any) vector.Any {
+	patternArg, inputArg := args[0], args[1]
+	var defArg vector.Any = vector.NewConst(super.NullString, args[0].Len(), nil)
```
Comment (nwt) [inline runtime/vam/expr/function/grok.go:29]: Nit:
```suggestion
	defArg := vector.Any(vector.NewConst(super.NullString, args[0].Len(), nil))
```

**Review by nwt — APPROVED** [review]

File: runtime/ztests/expr/function/grok.yaml (hunk 1 above)
Comment (mattnibs) [reply runtime/ztests/expr/function/grok.yaml:10]: What do you think they should do with null values? Is the current implementation incorrect?

Comment (mattnibs) [reply runtime/ztests/expr/function/grok.yaml:10]: I'll just have it return a null empty record when null inputs are encountered. Seems consistent.
